File: sort/python/src/sort/bubbleSort.py
# sort/python/src/sort/bubbleSort.py

from ._sort import swap

# ------------------------------------------------------------------------
#     bubblSort  -   bubble sort implementation
# ------------------------------------------------------------------------

def bubbleSort(toSort: list, reverse=False) -> list:
    """Bubble sort function.

    :param toSort: array to be sorted
    :type toSort: list
    :param reverse: sort array highest to lowest, defaults to False
    :type reverse: bool, optional
    :return: sorted array
    :rtype: list
    """        
    ar = toSort.copy()

    # iterate through array
    i = 0;
    while i < len(ar) - 1:
        if (not reverse and ar[i] > ar[i + 1]) or (reverse and ar[i] < ar[i + 1]):
            swap(ar, i, i + 1)
            i -= 1
        else:
            i += 1
        if i < 0:
            i = 0

    return ar

File: sort/c/src/sort/bubbleSort.c
// sort/c/src/bubbleSort.c

#include <stdio.h>
#include <stdbool.h>
#include "sort.h"

/*
------------------------------------------------------------------------
    bubblSort  -   bubble sort implementation
------------------------------------------------------------------------
*/

// bubble sort function
void bubbleSort(array *arptr, bool rev) {
    // iterate through array
    if (arptr->len > 1){
        int i = 0;
        while (i < arptr->len - 1) {
            if ((!rev && arptr->ar[i] > arptr->ar[i + 1]) || (rev && arptr->ar[i] < arptr->ar[i + 1])) {
                swap(arptr, i, i + 1);
                i -= 1;
            } else {
                i += 1;
            }
            if (i < 0) {
                i = 0;
            }
        }
    }
}